### PR TITLE
[Static if] replace overload constraints with static if (searching.d)

### PR DESCRIPTION
**Pull request by wilzbach**
Split up of #5145

### CODE REVIEW

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -1080,40 +1080,38 @@ if (isBidirectionalRange!Range && Needles.length > 1 &&
 
 /// Ditto
 bool endsWith(alias pred = "a == b", R1, R2)(R1 doesThisEnd, R2 withThis)
-if (isBidirectionalRange!R1 &&
-    isBidirectionalRange!R2 &&
-    is(typeof(binaryFun!pred(doesThisEnd.back, withThis.back)) : bool))
+if (isBidirectionalRange!R1 && (
+        is(typeof(binaryFun!pred(doesThisEnd.back, withThis)) : bool) ||
+        isBidirectionalRange!R2 && (is(typeof(binaryFun!pred(doesThisEnd.back, withThis.back)) : bool))
+    ))
```
Comment (wilzbach) [inline std/algorithm/searching.d]: The next overload is `isBidirectionalRange!R && is(typeof(binaryFun!pred(doesThisEnd.back, withThis)) : bool))`
-> `isBidirectionalRange!R` is the overlapping constraint.

**Comment by UplinkCoder** [comment]
The reason to have the overloads is to have the acceptable overloads visible.

**Comment by wilzbach** [comment]
> The reason to have the overloads is to have the acceptable overloads visible.

And the reason for unifying them under one function is that @WalterBright and others agree that implementation details shouldn't matter for the user.

**Comment by wilzbach** [comment]
Mostly superseded by https://github.com/dlang/phobos/pull/5575

**Pull request closed without merging**